### PR TITLE
A full editor can be used as git commit message editor

### LINKED ISSUE #30562: Git: Support editing the commit message in a text editor
When I make a commit, I enjoy writing the commit message in an editor window containing a complete diff of my staged changes. I can do that from the terminal by running the command:

```
git commit --verbose
```

However, VSCode doesn't seem to support that commit mode out of the box. Perhaps it could be the fallback behavior if you run one of the `Git: Commit` commands and enter a blank commit message. (Currently, that fails silently.) Or better yet, how about a `git.verboseCommit` setting that, set to `true`, would make the `Git: Commit` commands take me directly to the verbose commit editor?

### PR DESCRIPTION
**Pull request by JohnnyCrazy**
This PR fixes #30562

This PR introduces the ability to use a full editor pane to write a commit message.

https://user-images.githubusercontent.com/3109892/135883632-d8897493-0d66-40ca-bbf8-216ee411fb45.mp4

#### `git.useEditorToCommit: boolean`

If enabled, a full editor will be created for every commit action. These include triggered git commands like `Git: Commit All` and also `CTRL+ENTER` on an empty input box in the repository pane. If the input box is not empty, the full editor will **not** be used. This still allows for quick, small commit messages.

Once the editor is created, it behaves like `GIT_EDITOR="code --wait" git commit`: As soon as the file is closed, the contents will be used as message and the commit continues. If it's empty, the commit aborts and a `Aborting commit due to empty commit message` warning is shown.

#### `git.verboseCommit: boolean`

If enabled, the opened `COMMIT_EDITMSG` will contain more output. It's basically appending `--verbose` to the git call.

If `git.verboseCommit` is enabled but `git.useEditorToCommit` is not, a warning will be shown as the setting is pointless.

### Implementation

The implementation of the git commit integration is similar to the askpass one:

* We use the already existing IPC Server of the git extension
* Once a git commit is triggered, a custom `.sh/.bat` file is passed as the `GIT_EDITOR` environment variable to the git commit call
* This file starts a JS Process (no electron overhead) and passes all its arguments (Which is just the path to the `COMMIT_EDITMSG`) via IPC to the main electron process and waits for its answer.
* The main electron process opens the file and waits until its closed. If its closed, a dummy response is sent to the client, which terminates itself and thus, finishes the git commit process.

`vscode: git commit` waits for `git: commit` waits for `ipc-client-js: GIT_EDITOR` waits for `vscode: close of file`

### Testing

Testing was successful on windows and linux. I wanted to test the WSL Integration as well, but was unable to get it working with the downloaded VSIX in Code-OSS.

Initially I had a problem with the timeout of the IPC. This was solved by removing the timeout. This will also be [the default in node 13](https://github.com/nodejs/node/pull/27558)

### CODE REVIEW

File: extensions/git/src/commands.ts [hunk 1]
```diff
@@ -1397,6 +1397,19 @@ export class CommandCenter {
...
+		if (config.get<boolean>('useEditorToCommit')) {
+			opts.useEditor = true;
+		}
+
+		if (config.get<boolean>('verboseCommit')) {
+			opts.verbose = true;
+			if (!opts.useEditor) {
+				window.showInformationMessage(
+					localize('useless verbose commit', "Verbose committing has no effect. The setting 'git.useEditorToCommit' is not enabled.")
+				);
```
Comment (JohnnyCrazy) [inline extensions/git/src/commands.ts:1409]: I was not sure whether to include this warning or not. Up to you guys :)

File: extensions/git/src/commands.ts [hunk 2]
```diff
@@ -1446,10 +1459,13 @@ export class CommandCenter {
...
 	private async commitWithAnyInput(repository: Repository, opts?: CommitOptions): Promise<void> {
 		const message = repository.inputBox.value;
+		const root = Uri.file(repository.root);
+		const config = workspace.getConfiguration('git', root);
+
 		const getCommitMessage = async () => {
 			let _message: string | undefined = message;
 
-			if (!_message) {
+			if (!_message && !config.get<boolean>('useEditorToCommit')) {
```
Comment (JohnnyCrazy) [inline extensions/git/src/commands.ts:1468]: Without this additional check,

File: extensions/git/src/git.ts [hunk 3]
```diff
@@ -1300,8 +1302,16 @@ export class Repository {
 		}
 	}
 
-	async commit(message: string, opts: CommitOptions = Object.create(null)): Promise<void> {
-		const args = ['commit', '--quiet', '--allow-empty-message', '--file', '-'];
+	async commit(message?: string, opts: CommitOptions = Object.create(null)): Promise<void> {
+		const args = ['commit', '--quiet'];
+		const options: SpawnOptions = {};
+
+		if (!opts.useEditor || message) {
```
Comment (JohnnyCrazy) [inline extensions/git/src/git.ts:1309]: If a message is supplied, it should still be used for the commit. This allows to use the input box for quick commits, although `git.useEditorToCommit` is enabled.

File: extensions/git/src/gitEditor.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	async handle({ commitMessagePath }: GitEditorRequest): Promise<any> {
+		if (commitMessagePath) {
+			const file = Uri.file(commitMessagePath);
+			const doc = await workspace.openTextDocument(file);
+			const editor = await window.showTextDocument(doc);
+
+			// We don't want to remember the cursor position.
+			// One should be able to start writing the message immediately.
+			const position = new Position(0, 0);
+			editor.selection = new Selection(position, position);
```
Comment (JohnnyCrazy) [inline extensions/git/src/gitEditor.ts:47]: vscode seems to remember where my cursor was in an older file. This was quite annoying since I couldn't start writing the commit message immediately.

File: extensions/git/src/gitEditor.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+			return new Promise((c, e) => {
+				const onDidChange = window.onDidChangeVisibleTextEditors((editors) => {
+					if (editors.indexOf(editor) < 0) {
+						onDidChange.dispose();
+						return c(true);
+					}
+				});
+			});
```
Comment (JohnnyCrazy) [inline extensions/git/src/gitEditor.ts:56]: I'm not 100% sure about this part. Could be, that I missed some edge cases where this fails?

**Comment by JohnnyCrazy** [comment]
This is also particularly interesting for #2718 and #93150.

**Comment by JohnnyCrazy** [comment]
One test in windows is failing, seems to be not connected to my changes tho

**Comment by joshk0** [comment]
@JohnnyCrazy How can I help with this PR? I see that there are some merge conflicts. I could try to help you clean that up with a rebased version or a merge commit. Thanks for doing this!

File: extensions/git/src/gitEditor.ts (hunk 4 above)
Comment (joshk0) [reply extensions/git/src/gitEditor.ts:47]: This is probably because the path to the git-commit is always `.git/COMMIT_EDITMSG` or something. If you change how vscode calls git commit, you can make this a randomly named temp file. The downside is that you will lose the automated report that git commit gives you (especially when using `--verbose`) but then again, you have a whole IDE that can provide this information so maybe that's a good future path.

**Comment by JohnnyCrazy** [comment]
> 
> 
> @JohnnyCrazy How can I help with this PR? I see that there are some merge conflicts. I could try to help you clean that up with a rebased version or a merge commit. Thanks for doing this!

Thanks for the ping, missed that there are new changes and conflicts 👍 

I will have a look today/tomorrow and will report, if I run into any issues. Most changes seem to be straightforward to fix.

**Comment by joaomoreno** [comment]
No worries about the conflicts @JohnnyCrazy, I can take care of them.

Assigning this to May.

**Comment by joaomoreno** [comment]
Fantastic work here.

But because git is an extension, the way we handle the "closing of the COMMIT_MESSAGE" file is problematic:

- It's very easy for the user to simply opening another tab, thus cancelling (or worse, accepting) the commit operation
- In that situation, having an already opened `COMMIT_MESSAGE`, possibly dirty, will prevent further commits from being made

For that reason, I've brought these changes into [`joao/git-editor`](https://github.com/microsoft/vscode/tree/joao/git-editor) and will see if I can improve this using the FS Provider API.

Everything else works to perfection.

We might need to realign these settings and simply consider having this turned on by default.

File: extensions/git/src/gitEditor.ts (hunk 4 above)
Comment (Khady) [reply extensions/git/src/gitEditor.ts:47]: > This is probably because the path to the git-commit is always .git/COMMIT_EDITMSG or something. If you change how vscode calls git commit, you can make this a randomly named temp file. 

does it mean that not using .git/COMMIT_EDITMSG means no proper integration with prepare-commit-msg hook? it would be nice to preserve this

Comment (JohnnyCrazy) [reply extensions/git/src/gitEditor.ts:47]: Yes, that's why I refrained from using a custom/tmp file. I think the overall experience is still the best when the original file is used.

**Comment by JohnnyCrazy** [comment]
Yes, I already had a weird feeling when I was writing that "on close" part 😉 

Regarding the settings, one thing which held me back to make it the default action was the command trigger `Git: Commit ...`. I was unsure if some people might find the input box better than a whole file and don't like the change in the process.

Anyway, if you need anything or want me to further explore solutions using the FS Provider API, feel free to ping me!

**Comment by joaomoreno** [comment]
> Anyway, if you need anything or want me to further explore solutions using the FS Provider API, feel free to ping me!

Actually, yeah, if you'd like to give it a go, here's what I was thinking.

We can't really figure out when the file has been closed or not, this is a long standing open feature request into our API which won't get resolved any time soon. So we're going to have to go with the same semantics as you currently have. But in doing so, we need to use a different file on for each commit, so we avoid having potential save conflicts. Idea:

When we get pinged to open that `COMMIT_MESSAGE` file, instead of directly opening it we copy its contents into an in-memory virtual file, coming in from the [git fs provider](https://github.com/microsoft/vscode/blob/709a3508b2458b37ef03e00847b3678e0dad0092/extensions/git/src/fileSystemProvider.ts#L36-L36) (or possibly a new fs provider, so we don't clash with the existing one). We can have different files by doing something like `gitcommit:/<generated id>/COMMIT_MESSAGE`, that way the file would still have colorization etc. And we can simply delete the file once the commit operation is done.

What do you think?

**Comment by JohnnyCrazy** [comment]
> We can't really figure out when the file has been closed or not, this is a long standing open feature request into our API which won't get resolved any time soon. So we're going to have to go with the same semantics as you currently have. But in doing so, we need to use a different file on for each commit, so we avoid having potential save conflicts. Idea:

> When we get pinged to open that COMMIT_MESSAGE file, instead of directly opening it we copy its contents into an in-memory virtual file, coming in from the git fs provider (or possibly a new fs provider, so we don't clash with the existing one). We can have different files by doing something like gitcommit:/<generated id>/COMMIT_MESSAGE, that way the file would still have colorization etc. And we can simply delete the file once the commit operation is done.

Sounds good, the only question I currently have is regarding `So we're going to have to go with the same semantics as you currently have`: 

So, as I understood, we will continue to use `window.onDidChangeVisibleTextEditors`, but not with the real file but with a in-memory mapped file provided by a FS Provider. And if for some reason the commit aborts/continues while the file is still open, we simply "delete" the in-memory mapped file and the user is unable to edit the file anymore.

**Comment by joaomoreno** [comment]
> So, as I understood, we will continue to use `window.onDidChangeVisibleTextEditors`, but not with the real file but with a in-memory mapped file provided by a FS Provider. And if for some reason the commit aborts/continues while the file is still open, we simply "delete" the in-memory mapped file and the user is unable to edit the file anymore, even if it is open in an editor. (it probably shows the `(deleted)` suffix).

Absolutely right! It's definitely better than the current approach. And we can discuss with @jrieken and @bpasero if we can't have the notion of _editors which should automatically close if their document gets deleted_.

**Comment by joaomoreno** [comment]
@JohnnyCrazy Just resolved the conflicts. Looking forward to the further changes! 💪

**Comment by joaomoreno** [comment]
Open Editors API could be coming this month: https://github.com/microsoft/vscode/issues/15178#issuecomment-624512992! 🔥 At least in some proposed state, which we could use in git. Let's keep an eye on that, it would be very helpful here. 

I wonder if, given that API, we should still do the FS Provider dance. 🤔 My gut feeling tells me we don't actually have to go with that hassle, especially because git can't handle multiple simultaneous commits. So I propose we wait for that proposed API to advance and experiment with that, what do you think @JohnnyCrazy?

**Comment by JohnnyCrazy** [comment]
Yes, I agree. In the end we would abuse a completely different feature for our purposes and the code gets more complicated.

Looking forward to an Editor Tab API! 👍

**Comment by JohnnyCrazy** [comment]
As far as I understood #15178, it will take some time.

I would happily invest some more time finding a solution for the whole open editor problem, maybe involving the FS Provider API like you suggested. What is your take on this @joaomoreno ? I could imagine once the Editor API is out, the refactoring amount is rather minimal. But in the end you have more experience here, so up to you 👍

**Comment by joaomoreno** [comment]
#15178 is definitely in the horizon... we're just a bit overloaded and other stuff got prioritized.

We can definitely see if the FS Provider gives us a better experience here, so we're not blocked by that feature request. Let me know how it goes.

**Comment by zpanderson** [comment]
Just curious if this would get merged any time soon, we'd love to start using this in our process.

**Comment by JohnnyCrazy** [comment]
@joaomoreno I pushed a first implementation of a fs-provider based solution in f18ba9a0bf6ef2bd299b9b8548b09627f2b51912

Some notes for that:

Dirty commit message files are not a problem anymore. We're handling all the contents in-memory and the real file is only touched 2 times: When starting the commit and when it finishes up. Using an uuid as the URI `authority` make it also unique.

> When we get pinged to open that COMMIT_MESSAGE file, instead of directly opening it we copy its contents into an in-memory virtual file, coming in from the git fs provider (or possibly a new fs provider, so we don't clash with the existing one). We can have different files by doing something like gitcommit:/<generated id>/COMMIT_MESSAGE

Right now, the URI (incl. uuid) for the file is generated when we receive the IPC call. Thus, we can only delete the in-memory file once the callback to `onDidChangeVisibleTextEditors` finishes. The ideal solution would be to delete the file when the commit operation fails/succeeds. This would require generating the URI/uuid already in the `git.commit` call and passing it to the `GIT_EDITOR` proxy script. I think the only way to do correctly do this is via a env variable. What do you think of that? This would also mean `git.ts` using the `workspace.fs` api.


Instead of directly comparing if the spawned `editor` is still open, I changed it to compare based on the `URI`. This means if you split/duplicate the opened COMMIT_MSG and close the original editor, the commit is still in progress and only finishes if no editor with the corresponding URI is visible.


The problem with opening a new tab of course still exists due to `onDidChangeVisibleTextEditors`. The more I think about it, the more I think this is a blocker. Imagine writing a commit message and in the process of it, you want to lookup something in the code. You hit save and open a new tab --> The commit goes through with a unfinished message. Not transparent and not clear to the user.

**Comment by Khady** [comment]
I don't know if you are aware of https://github.com/kahole/edamagit but it solves the problem tackled by this PR. Maybe it doesn't cover all the edge cases. But it is a very good solution in the meantime.

**Comment by ashthespy** [comment]
Friendly check to ask if there has been any progress on this?
This is one of the biggest pain points while trying to migrate from Atom to VSCode. Being able to see a list of files in this commit while writing the commit message is invaluable in my work flow.

**Comment by JohnnyCrazy** [comment]
From my side, this is still blocked by #15178 and I don't think a viable alternative exists.

**Comment by ashthespy** [comment]
> From my side, this is still blocked by #15178 and I don't think a viable alternative exists.

In my limited understanding, the main blocker is handling such edge cases? 

> The problem with opening a new tab of course still exists due to `onDidChangeVisibleTextEditors`. The more I think about it, the more I think this is a blocker. Imagine writing a commit message and in the process of it, you want to lookup something in the code. You hit save and open a new tab --> The commit goes through with a unfinished message. Not transparent and not clear to the user.

If this is the only blocker, given years of everyone knowing saving/closing the `COMMIT_EDITMSG` message commits it, I'd reckon this is quite an edge case? Would it be possible to utilise your implementation now, and improve it when #15178 finally lands?

**Comment by rdnlsmith** [comment]
@ashthespy

> This is one of the biggest pain points while trying to migrate from Atom to VSCode.

I've been using a user task to accomplish this for a few months now (writeup [here](https://rdnlsmith.com/posts/a-better-way-to-write-commit-messages-in-vs-code/)). I previously [commented in the issue](https://github.com/microsoft/vscode/issues/30562#issuecomment-597296492) with the basis of this idea, but I hadn't figured out how to set the environment variables when on Windows (fixed now). I'll celebrate the day I can delete my tasks, but maybe this can tide you over in the meantime!

**Comment by JohnnyCrazy** [comment]
> If this is the only blocker, given years of everyone knowing saving/closing the `COMMIT_EDITMSG` message commits it, I'd reckon this is quite an edge case? Would it be possible to utilise your implementation now, and improve it when #15178 finally lands?

The blocker is not the saving/closing of the file, This indeed works like intended and is the standard in other editors as well. The problem is, that when you switch tabs, the `COMMIT_MSG` will be submitted as well and this very unnatural and not transparent behaviour to the user.

**Comment by ashthespy** [comment]
> The problem is, that when you switch tabs, the `COMMIT_MSG` will be submitted as well and this very unnatural and not transparent behaviour to the user.

Ouch yes. That would indeed be annoying. Anything we can do to help/push #15178 along?

I just ran into an issue with @rdnlsmith workaround as it defaults to the first folder/repo in the workspace for some reason

**Comment by rdnlsmith** [comment]
@ashthespy Ah, sorry—I've only used single-repo workspaces before. Perhaps you could use an input to select which repo you want, and use that with the `--git-dir` option? This would be similar to the last example in my post, for the "commit as different author" task.

I don't want to clutter up this PR thread too much, but feel free to contact me directly if you have trouble with this. My email is in my profile.

**Comment by JohnnyCrazy** [comment]
Hi,

With the new proposed tabs API (#133532), I updated the branch to better support detecting if an editor with the commit message is still open ([source](https://github.com/microsoft/vscode/pull/95266/files#diff-7dd9d61f75a4eace2abe9421819b3f5e5bbdf685490a6e7621f038bb1b7cc4ffR34)) and it works flawlessly so far!



https://user-images.githubusercontent.com/3109892/135883632-d8897493-0d66-40ca-bbf8-216ee411fb45.mp4


One open question is the general structure and error handling of `commitFileSystemProvider.ts`. It seems like the assignment chain ended with you @lszomoru, do you have any thoughts about this? Should we add better error handling?

**Comment by joaomoreno** [comment]
@JohnnyCrazy That's pretty cool! Looking forward to getting this merged once that API lands.

**Comment by leonardopsantos** [comment]
Any news on this?

**Comment by lszomoru** [comment]
@lramos15, how close is the tabs API from being finalised?
Is it in a state that we can make progress on this pull request?

**Comment by lramos15** [comment]
> @lramos15, how close is the tabs API from being finalised? Is it in a state that we can make progress on this pull request?

@lszomoru We're currently on track to finalize for this milestone, but it may be a partial finalization i.e. reading only and not close, move, or any layout related information as we were looking into the possibility of the tabs API becoming a full layout manager and don't want to back ourselves into a corner.

**Comment by lszomoru** [comment]
@joaomoreno, this PR is not ready to be reviewed. Let me know if you have any questions or concerns. Thanks!

**Comment by rmunn** [comment]
> @joaomoreno, this PR is not ready to be reviewed. Let me know if you have any questions or concerns. Thanks!

@lszomoru - Did you mean "not" ready? Or "now" ready?

**Comment by lszomoru** [comment]
@rmunn, thanks! Updated my comment. This is **now** ready to be reviewed.

File: src/vs/workbench/contrib/scm/browser/util.ts [hunk 6]
```diff
@@ -37,7 +37,7 @@ export function isSCMResource(element: any): element is ISCMResource {
 	return !!(element as ISCMResource).sourceUri && isSCMResourceGroup((element as ISCMResource).resourceGroup);
 }
 
-const compareActions = (a: IAction, b: IAction) => a.id === b.id;
+const compareActions = (a: IAction, b: IAction) => a.id === b.id && a.enabled === b.enabled;
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/browser/util.ts:40]: Interesting!

**Review by joaomoreno — APPROVED** [review]

File: src/vs/workbench/contrib/scm/browser/util.ts (hunk 6 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/scm/browser/util.ts:40]: As discussed, this is necessary in order for the enablement to be refreshed in the repository toolbar.

File: extensions/git/src/gitEditor/gitEditor.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+	private disposable: IDisposable = EmptyDisposable;
+
+	constructor(private ipc?: IIPCServer) {
+		if (ipc) {
+			this.disposable = ipc.registerHandler('git-editor', this);
+		}
+	}
+
+	async handle({ commitMessagePath }: GitEditorRequest): Promise<any> {
```
[the remark this answers is not in the document]
Comment (lszomoru) [reply extensions/git/src/gitEditor/gitEditor.ts:24]: Makes sense. I will follow up with a commit to fix that.

**Comment by joaomoreno** [comment]
This is awesome. Great job @lszomoru for getting this done! And many thanks to @JohnnyCrazy for being so patient with us and collaborating with us along the way! 🚀 👏

**Comment by lszomoru** [comment]
All the credit goes to @JohnnyCrazy 🚀

**Comment by JohnnyCrazy** [comment]
Good stuff @lszomoru ! Really enjoyed working on this one, thanks for the guidance and help along the way @joaomoreno ! 😊

**Comment by lszomoru** [comment]
@JohnnyCrazy, after we have merged the PR we ran into some issues and we had to revert the changes. Unfortunately, we ran out of time before we had to shut down the main branch for the release, so these changes did not make it into the May release. I have create a new PR (https://github.com/microsoft/vscode/pull/151491) to get these changes into the June release.
